include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge_sort.h

```diff
@@ -159,6 +310,72 @@ struct __parallel_sort_submitter<_IdType, __internal::__optional_kernel_name<_Le
     }
 };
 
+template <typename _Range, typename _Compare>
+struct __leaf_sorter_selector
+{
+    // 8 is the maximum reasonable value for bubble sub-group sorter due to algorithm complexity
+    // TODO: reconsider the value if another algorithm is used,
+    //       or an internal cap is set (e.g. sorting 2 sequences of 4/8 items each)
+    using _Leaf8 = __leaf_sorter<8, _Range, _Compare>;
+    using _Leaf4 = __leaf_sorter<4, _Range, _Compare>;
+    // 2 is the smallest reasonable value for merge-path group sorter since it loads 2 values at least
+    using _Leaf2 = __leaf_sorter<2, _Range, _Compare>;
+
+    using _Tp = oneapi::dpl::__internal::__value_t<_Range>;
+
+    std::variant<_Leaf8, _Leaf4, _Leaf2>
+    select(const sycl::queue& __q, _Range& __rng, _Compare __comp) const
+    {
+        const std::size_t __n = __rng.size();
+        auto&& __d = __q.get_device();
+
+        std::size_t __max_wg_size = __d.template get_info<sycl::info::device::max_work_group_size>();
+        __max_wg_size = oneapi::dpl::__internal::__dpl_bit_floor(__max_wg_size);
+
+        const auto __sg_sizes = __d.template get_info<sycl::info::device::sub_group_sizes>();
+        const auto __max_sg_size = __sg_sizes.empty() ? 1 : *std::max_element(__sg_sizes.begin(), __sg_sizes.end());
+        // __oversubscription is similar to "theoretical occupancy" in GPU, or "multithreading" in CPU
+        // TODO: reconsider the constant if the corresponding query appears in the SYCL specification
+        // 8 (or 6, which is slightly less) appears to be common for modern Intel/AMD/Nvidia GPUs see:
+        // Intel: https://www.intel.com/content/www/us/en/docs/oneapi/optimization-guide-gpu/2024-2/intel-xe-gpu-architecture.html:
+        //        see: "Threads / XVE" (8)
+        // Nvidia: https://xmartlabs.github.io/cuda-calculator (they deprecated their official calculator with no alternative):
+        //         see: "Active Thread Blocks per Multiprocessor" (6)
+        // AMD: https://rocm.docs.amd.com/en/latest/how-to/tuning-guides/mi300x/workload.html#mi300x-occupancy-vgpr-table
+        //         see: "Occupancy per EU" (8)
+        const std::uint8_t __oversubscription = __d.is_gpu() ? 8 : 1;
+        const auto __max_cu = __d.template get_info<sycl::info::device::max_compute_units>();
+        const auto __saturation_point = __max_cu * __max_sg_size * __oversubscription;
+        const auto __desired_data_per_workitem = __n / __saturation_point;
+
+        // Pessimistically double the memory requirement to take into account memory used by compiled kernel.
+        // TODO: investigate if the adjustment can be less conservative
+        const std::size_t __max_slm_items =
+            __d.template get_info<sycl::info::device::local_mem_size>() / (sizeof(_Tp) * 2);
+        if (__max_slm_items >= _Leaf8::storage_size(__max_wg_size) && __desired_data_per_workitem >= 8)
```

Comment:
May be better get this `8` number in condition `__desired_data_per_workitem >= 8` from `_Leaf8` struct - I am about `static constexpr std::uint16_t __data_per_workitem = _DataPerWorkitem;`

Reply:
At this place and below

Reply:
Would you mind if I leave it as is? Using just 8 looks simpler.

Reply:
ok